pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -538,6 +556,127 @@ public void deleteFailed(ManagedLedgerException exception, Object ctx) {
+            synchronized (dispatcher0) {
+                updatePendingAckMessagesAfterAcknowledged0(ackFrom, positions);
+            }
+        } else {
+            updatePendingAckMessagesAfterAcknowledged0(ackFrom, positions);
+        }
+    }
+
+    private void updatePendingAckMessagesAfterAcknowledged0(Consumer ackFrom, List<Position> positions) {
+        int attemptAckMsgs = 0;
+        for (Position position : positions) {
+            final long ledgerId = position.getLedgerId();
+            final long entryId = position.getEntryId();
+            final int batchMessagesAckedCount = AckSetStateUtil.getBatchMessagesAckedCount(position);
+            final boolean positionRemovedFromCursor = AckSetStateUtil.isPositionRemovedFromCursor(position);
+            if (batchMessagesAckedCount == 0) {
+                // All messages were skipped.
+                // Since we can not get how many msgs that were attempted to ack, just plus 1 into the
+                // "attemptAckMsgs".
+                attemptAckMsgs++;
+                log.warn("[{}][{}]{}-{}-{} is acknowledging {}:{}, which has been acked before. consumer_size: {}",
+                        topicName, subName,
+                        ackFrom == null ? "null" : ackFrom.cnx(),
+                        ackFrom == null ? "null" : ackFrom.consumerId(),
+                        ackFrom == null ? "null" : ackFrom.consumerName(),
+                        ledgerId, entryId, getConsumers().size());
+                continue;
+            }
+            // Find the messages' owner and update un-acknowledged messages.
+            Consumer owner = null;
+            IntIntPair batchSizeAndHashPair = ackFrom == null ? null
+                    : positionRemovedFromCursor
+                    ? ackFrom.getPendingAcks().removeAndReturn(ledgerId, entryId)
+                    : ackFrom.getPendingAcks().get(ledgerId, entryId);
+            if (batchSizeAndHashPair != null) {
+                owner = ackFrom;
+            } else {
+                for (Consumer consumer : getConsumers()) {
+                    if (consumer == ackFrom) {
+                        continue;
+                    }
+                    batchSizeAndHashPair = positionRemovedFromCursor
+                            ? consumer.getPendingAcks().removeAndReturn(ledgerId, entryId)
+                            : consumer.getPendingAcks().get(ledgerId, entryId);
+                    if (batchSizeAndHashPair != null) {
+                        // Continue find the owner
+                        owner = consumer;
+                        break;
+                    }
+                }
+            }
+            if (owner == null) {
+                // Since we can not get how many msgs that were attempted to ack, just plus 1 into the
+                // "attemptAckMsgs".
+                attemptAckMsgs++;
+                log.warn("[{}][{}]{}-{}-{} skipped to reduce un-ack-msgs for {}:{}, because could not find the"
```

Comment:
Is warn level too high? Isn't the actual problem resolved by this PR and there's nothing to do about it?

Reply:
I think it is okay, this warning log indicates there are repeatable consumption

Reply:
I'd suggest changing from warn to info, similar rationale as in https://github.com/apache/pulsar/pull/24096#discussion_r2007094289.

Reply:
Improved